### PR TITLE
Fix: On the Web, cannot support multiline inputting when registering customized TextInputControl

### PR DESCRIPTION
**Pull request by Satsrag**
For [flutter/flutter/125875](https://github.com/flutter/flutter/issues/125875)

When registering customized TextInputControl, the _PlatformTextInputControl sends inputType = TextInputType.none to the engine. After receiving TextInputType.none, the engine on the Web will create a `<input>` element instead of  `<textarea>`. So there is no way to input \n(multiline).

This is my solution. I tested in Android Chrome, iOS Safari, and macOS Chrome. It works for me. But I'm not sure about other use cases. I pleasure, If someone gives me suggestions.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text_editing/input_type.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing/input_type.dart:97]: So `TextInputType.none` is always treated as multiline? I don't think that's correct. cc @Renzo-Olivares for ideas on how to distinguish between single-line and multi-line modes when using `TextInputType.none`.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1441]: If we end up doing this, we will need a test to make sure it doesn't break in the future. You can take a look here for an example: https://github.com/flutter/engine/blob/4a14ee8072a3d3b12b830ca996fd656a6bca774b/lib/web_ui/test/engine/text_editing_test.dart#L421

**Review by mdebbar — COMMENTED** [review]
Thanks for the contribution!

It makes sense to fix this bug, but I'm not sure if we should always treat `TextInputType.none` as multiline.

File: lib/web_ui/lib/src/engine/text_editing/input_type.dart
Comment (htoor3) [reply lib/web_ui/lib/src/engine/text_editing/input_type.dart:97]: Agreed here.  What's the reasoning behind setting `none` as an input type for [custom text input controls](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/text_input.dart#L2229) in the framework?  Could we just make the determination at the framework level (via `maxLines > 1` or something similar) and just send an explicit input type to the engine?

Comment (Satsrag) [reply lib/web_ui/lib/src/engine/text_editing/input_type.dart:97]: I agreed to solve via framework level.
However, I think, the main reason for sending `none` as an input type is to disable the soft keyboard display. 
So, we may add a property (such as disableSoftKeyboard) to the class `InputConfiguration`. Then we will add code to all platforms to handle this property to disable soft keyboard display.

Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/input_type.dart:97]: Agreeing with @Satsrag, I think we need to continue setting the `inputMode` of the web dom element (input/textarea) to `none` to block the virtual keyboard when a custom text input control is provided. This means we can't conditionally send an explicit input type based on `maxLines` (unless we add more input types).

Adding a flag to `TextInputConfiguration` as suggested by @Satsrag sounds reasonable, but `disableSoftKeyboard` sounds like it would accomplish the same thing as passing [`TextInputType.none`](https://api.flutter.dev/flutter/services/TextInputType/none-constant.html). Maybe what we need is a `forceMultiline/isMultiline` parameter which sets its value based on the `maxLines` field. Then the web engine can use this to decide whether to use the  `input/textarea`.

Comment (Satsrag) [reply lib/web_ui/lib/src/engine/text_editing/input_type.dart:97]: Yes, I agree with @Renzo-Olivares.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (Satsrag) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart]: Added test for `EngineInputType.multilineNone`. Is there a need for more tests? Thank you.

File: lib/web_ui/lib/src/engine/text_editing/input_type.dart
Comment (htoor3) [inline lib/web_ui/lib/src/engine/text_editing/input_type.dart:100]: We should probably have a clear comment description of why this is needed or in what case this type might be used

File: lib/web_ui/lib/src/engine/text_editing/input_type.dart
Comment (Satsrag) [inline lib/web_ui/lib/src/engine/text_editing/input_type.dart:113]: @htoor3 
Is the comment clear now? I'm sorry, I'm not good at English. If it's still unclear, could you please write it for me to copy? Thank you.

Comment (htoor3) [reply lib/web_ui/lib/src/engine/text_editing/input_type.dart:113]: Thank you @Satsrag! It looks good to me, I would just add another line that points to the issue this is fixing like
```
/// See: https://github.com/flutter/flutter/issues/125875
```

**Comment by Satsrag** [comment]
I'm sorry about asking for review again and again. But this time, I think that all work is done. Ask for a review again. 
In [framework PR](https://github.com/flutter/flutter/pull/140356), `forceMultiline` was renamed to `isMultiline`. So here I was renamed to `isMultiline` and added tests.

File: lib/web_ui/lib/src/engine/text_editing/input_type.dart
Comment (htoor3) [inline lib/web_ui/lib/src/engine/text_editing/input_type.dart:17]: nit: change `isMultiline= false` to `isMultiline = false`

**Review by htoor3 — APPROVED** [review]
Thank you for this! This lgtm.  Before merging, let's verify the framework approach is the correct one with @Renzo-Olivares

File: lib/web_ui/lib/src/engine/text_editing/input_type.dart
Comment (Satsrag) [reply lib/web_ui/lib/src/engine/text_editing/input_type.dart]: Sorry about this. Now I have fixed it. Thank you!

Comment (Satsrag) [reply lib/web_ui/lib/src/engine/text_editing/input_type.dart]: @htoor3, Happy New Year! [The framework PR](https://github.com/flutter/flutter/pull/140356) has been merged. Is it time to merge this PR?

**Comment by htoor3** [comment]
lgtm, thank you @Satsrag for your contribution!

**Comment by Satsrag** [comment]
@mdebbar, Please have a review.

**Comment by Satsrag** [comment]
@htoor3 when this PR will be merged?

**Comment by htoor3** [comment]
Let's get a review from @mdebbar and then we can get this merged

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1286]: Small suggestion to simplify this a bit:
```suggestion
    if (config.inputType.inputmodeAttribute == 'none') {
      activeDomElement.setAttribute('inputmode', 'none');
    }
```

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1466]: Small suggestion to simplify this a bit:
```suggestion
        if (inputConfiguration.inputType is! MultilineInputType) {
```
and make it `class MultilineNoTextInputType extends MultilineInputType { ... }`.

**Review by mdebbar — APPROVED** [review]
Thanks for helping make Flutter better!

The PR looks good to me with some minor suggestions.

**Comment by Satsrag** [comment]
> Thanks for helping make Flutter better!
> 
> The PR looks good to me with some minor suggestions.

@mdebbar Thank you! Fixed according to your suggestion.

**Pull request merged**
